### DISCUSSION TITLE
Tesla Charging

### DISCUSSION THREAD

**Discussion by brianmay** [amberelectric/public-api#128] [General]
I have been writing software to automatically charge my Tesla
automatically based on the current amber prices. The challenge is that
that the cheapest time is often during the day, but the Tesla may not be
available then. So I still need to charge overnight also.

I currently have the following rules:

* charge to 50% always.
* charge to 70% at 2am.
* charge to 80% if price < 15c
* charge to 90% if price < 10c

The problem I am encountering is that the current price seems to
fluctuate a lot. For example right now the price is mostly 13c, but
every now and again it jumps up to 16c. Which means the car stops
charging when there is no need for it to stop charging.

Wondering if this is something to do with 5 minutes intervals vs 30
minute intervals. At the moment I am retrieving 30 minute interval data.
Would going to 5 minute help?

Any ideas or suggestions?

**Comment by madpilot** [comment]
The market naturally jumps around - using 5-minute intervals will actually make it more jumpy (30-minute prices are the average of the 5-minute price, so you get some level of smoothing).

Some things you could do:

1) taking the average of a large time frame: maybe the last two and the next two hours. This will smooth the transitions, but may mean you pay a little more
2) exponential filter: which lets you tune you filter to care more about the past (less movement, slower response) or more about the current (more movement, faster response).
3) low pass filter (only change if there has been and increase or decrease for a number of intervals in a row)
4) some sort of cost minimisation function (use linear algebra to find the local minima of the cost function over a period of time using forecasts)

Just be aware that forecasts get worst during periods of volatility.

Another option is to just not worry about it. If the price is higher for 30 minutes and you stop charging, it might be fine.

**Comment by brianmay** [comment]
Not sure if I was clear, my issue is that the price keeps jumping inside the current 30 minute interval. So it is not clear what the price will be until after it is too late to use it for that period.

Your solutions sound like maths. Arrghh. Runs away screaming :-)

Actually I guess the the best solution would be create several different models, estimate the costs for each one, and then pick the best one. Need to think about if I can be bothered or not...

A complication in my case is that prices rise considerably between 3pm and 9pm due to distributor costs, so any model that looks at past data would need to take that into account also.

**Reply by madpilot** [reply]
Ah. Right. Yeah, it jumps around because it is an estimate until the last 5 minutes of the interval (it uses forecast prices before that).

The five-minute price has the same problem.

Once we start billing in 5-minute intervals the problem… doesn't go away but at least the price will be accurate to your bill.
